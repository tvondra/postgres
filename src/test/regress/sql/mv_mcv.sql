-- data type passed by value
CREATE TABLE mcv_list (
    a INT,
    b INT,
    c INT
);

-- unknown column
ALTER TABLE mcv_list ADD STATISTICS (mcv) ON (unknown_column);

-- single column
ALTER TABLE mcv_list ADD STATISTICS (mcv) ON (a);

-- single column, duplicated
ALTER TABLE mcv_list ADD STATISTICS (mcv) ON (a, a);

-- two columns, one duplicated
ALTER TABLE mcv_list ADD STATISTICS (mcv) ON (a, a, b);

-- unknown option
ALTER TABLE mcv_list ADD STATISTICS (unknown_option) ON (a, b, c);

-- missing MCV statistics
ALTER TABLE mcv_list ADD STATISTICS (dependencies, max_mcv_items 200) ON (a, b, c);

-- invalid mcv_max_items value / too low
ALTER TABLE mcv_list ADD STATISTICS (mcv, max_mcv_items 10) ON (a, b, c);

-- invalid mcv_max_items value / too high
ALTER TABLE mcv_list ADD STATISTICS (mcv, max_mcv_items 10000) ON (a, b, c);

-- correct command
ALTER TABLE mcv_list ADD STATISTICS (mcv) ON (a, b, c);

-- random data
INSERT INTO mcv_list
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);

ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

TRUNCATE mcv_list;

-- a => b, a => c, b => c
INSERT INTO mcv_list
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);

ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

TRUNCATE mcv_list;

-- a => b, a => c
INSERT INTO mcv_list
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);

ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

TRUNCATE mcv_list;

-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO mcv_list
     SELECT i/10000, i/20000, i/40000 FROM generate_series(1,1000000) s(i);
CREATE INDEX mcv_idx ON mcv_list (a, b);

ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

EXPLAIN (COSTS off)
 SELECT * FROM mcv_list WHERE a = 10 AND b = 5;

DROP TABLE mcv_list;

-- varlena type (text)
CREATE TABLE mcv_list (
    a TEXT,
    b TEXT,
    c TEXT
);

ALTER TABLE mcv_list ADD STATISTICS (mcv) ON (a, b, c);

-- random data
INSERT INTO mcv_list
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);

ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

TRUNCATE mcv_list;

-- a => b, a => c, b => c
INSERT INTO mcv_list
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);

ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

TRUNCATE mcv_list;

-- a => b, a => c
INSERT INTO mcv_list
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

TRUNCATE mcv_list;

-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO mcv_list
     SELECT i/10000, i/20000, i/40000 FROM generate_series(1,1000000) s(i);
CREATE INDEX mcv_idx ON mcv_list (a, b);
ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

EXPLAIN (COSTS off)
 SELECT * FROM mcv_list WHERE a = '10' AND b = '5';

TRUNCATE mcv_list;

-- check explain (expect bitmap index scan, not plain index scan) with NULLs
INSERT INTO mcv_list
     SELECT
       (CASE WHEN i/10000 = 0 THEN NULL ELSE i/10000 END),
       (CASE WHEN i/20000 = 0 THEN NULL ELSE i/20000 END),
       (CASE WHEN i/40000 = 0 THEN NULL ELSE i/40000 END)
     FROM generate_series(1,1000000) s(i);
ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

EXPLAIN (COSTS off)
 SELECT * FROM mcv_list WHERE a IS NULL AND b IS NULL;

DROP TABLE mcv_list;

-- NULL values (mix of int and text columns)
CREATE TABLE mcv_list (
    a INT,
    b TEXT,
    c INT,
    d TEXT
);

ALTER TABLE mcv_list ADD STATISTICS (mcv) ON (a, b, c, d);

INSERT INTO mcv_list
     SELECT
         mod(i, 100),
         (CASE WHEN mod(i, 200) = 0 THEN NULL ELSE mod(i,200) END),
         mod(i, 400),
         (CASE WHEN mod(i, 300) = 0 THEN NULL ELSE mod(i,600) END)
     FROM generate_series(1,10000) s(i);

ANALYZE mcv_list;

SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;

DROP TABLE mcv_list;
