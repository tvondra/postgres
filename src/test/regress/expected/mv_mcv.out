-- data type passed by value
CREATE TABLE mcv_list (
    a INT,
    b INT,
    c INT
);
-- unknown column
CREATE STATISTICS s4 WITH (mcv) ON (unknown_column) FROM mcv_list;
ERROR:  column "unknown_column" referenced in statistics does not exist
-- single column
CREATE STATISTICS s4 WITH (mcv) ON (a) FROM mcv_list;
ERROR:  statistics require at least 2 columns
-- single column, duplicated
CREATE STATISTICS s4 WITH (mcv) ON (a, a) FROM mcv_list;
ERROR:  duplicate column name in statistics definition
-- two columns, one duplicated
CREATE STATISTICS s4 WITH (mcv) ON (a, a, b) FROM mcv_list;
ERROR:  duplicate column name in statistics definition
-- unknown option
CREATE STATISTICS s4 WITH (unknown_option) ON (a, b, c) FROM mcv_list;
ERROR:  unrecognized STATISTICS option "unknown_option"
-- correct command
CREATE STATISTICS s4 WITH (mcv) ON (a, b, c) FROM mcv_list;
-- random data
INSERT INTO mcv_list
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | f         | 
(1 row)

TRUNCATE mcv_list;
-- a => b, a => c, b => c
INSERT INTO mcv_list
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | t         | nitems=1000
(1 row)

TRUNCATE mcv_list;
-- a => b, a => c
INSERT INTO mcv_list
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | t         | nitems=1000
(1 row)

TRUNCATE mcv_list;
-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO mcv_list
     SELECT i/100, i/200, i/400 FROM generate_series(1,10000) s(i);
CREATE INDEX mcv_idx ON mcv_list (a, b);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | t         | nitems=100
(1 row)

EXPLAIN (COSTS off)
 SELECT * FROM mcv_list WHERE a = 10 AND b = 5;
                 QUERY PLAN                 
--------------------------------------------
 Bitmap Heap Scan on mcv_list
   Recheck Cond: ((a = 10) AND (b = 5))
   ->  Bitmap Index Scan on mcv_idx
         Index Cond: ((a = 10) AND (b = 5))
(4 rows)

DROP TABLE mcv_list;
-- varlena type (text)
CREATE TABLE mcv_list (
    a TEXT,
    b TEXT,
    c TEXT
);
CREATE STATISTICS s5 WITH (mcv) ON (a, b, c) FROM mcv_list;
-- random data
INSERT INTO mcv_list
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | f         | 
(1 row)

TRUNCATE mcv_list;
-- a => b, a => c, b => c
INSERT INTO mcv_list
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | t         | nitems=1000
(1 row)

TRUNCATE mcv_list;
-- a => b, a => c
INSERT INTO mcv_list
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | t         | nitems=1000
(1 row)

TRUNCATE mcv_list;
-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO mcv_list
     SELECT i/100, i/200, i/400 FROM generate_series(1,10000) s(i);
CREATE INDEX mcv_idx ON mcv_list (a, b);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | t         | nitems=100
(1 row)

EXPLAIN (COSTS off)
 SELECT * FROM mcv_list WHERE a = '10' AND b = '5';
                         QUERY PLAN                         
------------------------------------------------------------
 Bitmap Heap Scan on mcv_list
   Recheck Cond: ((a = '10'::text) AND (b = '5'::text))
   ->  Bitmap Index Scan on mcv_idx
         Index Cond: ((a = '10'::text) AND (b = '5'::text))
(4 rows)

TRUNCATE mcv_list;
-- check explain (expect bitmap index scan, not plain index scan) with NULLs
INSERT INTO mcv_list
     SELECT
       (CASE WHEN i/100 = 0 THEN NULL ELSE i/100 END),
       (CASE WHEN i/200 = 0 THEN NULL ELSE i/200 END),
       (CASE WHEN i/400 = 0 THEN NULL ELSE i/400 END)
     FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | t         | nitems=100
(1 row)

EXPLAIN (COSTS off)
 SELECT * FROM mcv_list WHERE a IS NULL AND b IS NULL;
                    QUERY PLAN                     
---------------------------------------------------
 Bitmap Heap Scan on mcv_list
   Recheck Cond: ((a IS NULL) AND (b IS NULL))
   ->  Bitmap Index Scan on mcv_idx
         Index Cond: ((a IS NULL) AND (b IS NULL))
(4 rows)

DROP TABLE mcv_list;
-- NULL values (mix of int and text columns)
CREATE TABLE mcv_list (
    a INT,
    b TEXT,
    c INT,
    d TEXT
);
CREATE STATISTICS s6 WITH (mcv) ON (a, b, c, d) FROM mcv_list;
INSERT INTO mcv_list
     SELECT
         mod(i, 100),
         (CASE WHEN mod(i, 200) = 0 THEN NULL ELSE mod(i,200) END),
         mod(i, 400),
         (CASE WHEN mod(i, 300) = 0 THEN NULL ELSE mod(i,600) END)
     FROM generate_series(1,10000) s(i);
ANALYZE mcv_list;
SELECT mcv_enabled, mcv_built, pg_mv_stats_mcvlist_info(stamcv)
  FROM pg_mv_statistic WHERE starelid = 'mcv_list'::regclass;
 mcv_enabled | mcv_built | pg_mv_stats_mcvlist_info 
-------------+-----------+--------------------------
 t           | t         | nitems=1200
(1 row)

DROP TABLE mcv_list;
